File: encryptionClasses/aes_cbc.py
from Crypto.Cipher import AES
from Crypto.Protocol.KDF import PBKDF2
from Crypto.Hash import HMAC, SHA256 
from Crypto import Random
from base64 import b64encode, b64decode


class AesCbc: 
	'''
	Encrypts and decrypts data using AES CBC 128 / 256.
	'''
	_sizes = (128, 256)
	_salt_size = 16
	_iv_size = 16
	_mac_size = 32
	
	def __init__(self, size=128): 
		'''
		:param size: int, the key size.
		:raises ValueError: if the size is invalid.
		'''
		if size not in self._sizes: 
			raise ValueError('Key size must be {} or {} bits.'.format(*self._sizes))
		self._key_size = size
		self.key = None
		self.rounds = 100000
		self.b64 = True
	
	def encrypt(self, data, password):
		'''
		Encrypts data (strings or bytes), returns a base64 encoded string or raw bytes. 
		
		:param data: string or bytes
		:param password: string
		:returns bytes or string (base64)
		'''
		data = self._pkcs_pad(data if type(data) is bytes else data.encode())
		iv, salt = self._ivgen(self._iv_size), self._ivgen(self._salt_size)
		self.key, mac_key = self._keygen(password, salt)
		
		cipher = AES.new(key=self.key, mode=AES.MODE_CBC, IV=iv)
		encrypted = cipher.encrypt(data)
		mac = self._sign(iv + encrypted, mac_key) 
		new_data = salt + iv + encrypted + mac
		
		if self.b64: 
			new_data = b64encode(new_data)
		return new_data
	
	def decrypt(self, data, password):
		'''
		Decrypts data - data could be bytes or string (base64 encoded bytes) 
		
		:param data: string or bytes
		:param password: string
		:returns bytes or None
		'''
		try:
			if type(data) is not bytes:
				data = data.encode()
			if self.b64: 
				data = b64decode(data)
			if len(data) < self._iv_size + self._salt_size + AES.block_size + self._mac_size: 
				raise ValueError("Not enough data.")
			
			salt, iv, encrypted, mac = (
				data[:self._iv_size], 
				data[self._iv_size:self._iv_size+self._salt_size], 
				data[self._iv_size+self._salt_size:-self._mac_size], 
				data[-self._mac_size:]
			)
			self.key, mac_key = self._keygen(password, salt)
			if not self._verify(iv + encrypted, mac, mac_key): 
				raise Exception("Verification failed.")
			
			cipher = AES.new(key=self.key, mode=AES.MODE_CBC, IV=iv)
			decrypted = cipher.decrypt(encrypted)
			return self._pkcs_unpad(decrypted)
		except Exception as e: 
			print(e)
	
	def _keygen(self, password, salt):
		'''
		Creates a pair of keys from the password. 
		
		:param password: string
		:param salt: bytes
		:returns tuple[bytes]
		'''
		key_size = int(self._key_size / 8)
		key = PBKDF2(password, salt, key_size * 2, self.rounds)
		return (key[:key_size], key[key_size:])
	
	def _ivgen(self, size=16): 
		'''
		Creates random bytes.
		
		:param size: int, optional
		:returns bytes
		'''
		return Random.new().read(size)
	
	def _sign(self, data, key): 
		'''
		Creates HMAC signature.
		
		:param data: bytes
		param key: bytes
		returns bytes
		'''
		hmac = HMAC.new(key, data, digestmod=SHA256)
		return hmac.digest()
	
	def _verify(self, data, mac, key): 
		'''
		Preforms HMAC verification.
		
		:param data: bytes
		:param mac: bytes
		:param key: bytes
		:returns bool
		'''
		return mac == self._sign(data, key)
	
	def _pkcs_pad(self, data): 
		'''Preforms PKCS7 padding.'''
		padding = AES.block_size - (len(data) % AES.block_size)
		return data + (chr(padding) * padding).encode()
	
	def _pkcs_unpad(self, data): 
		'''Removes PKCS7 padding.'''
		padding = ord(data[-1:])
		return data[:-padding]

File: encryptionClasses/AesCbc.cs
using System;
using System.Security.Cryptography;
using System.Text;
using System.IO;
using System.Collections.Generic;

/// <summary>
/// Encrypts and decrypts data using AES-CBC 128 / 256.
/// </summary>
class AesCbc
{
	private int[] sizes = new int[] { 128, 256 };
	private int ivSize = 16;
	private int macSize = 32;
	private int blockSize = 16;
	private int keySize = 128;
	public byte[] key = null;
	public int rounds = 100000;
	public bool b64 = true;

	/// <exception cref="ArgumentException">
	/// Thrown when invalid size is selected.
	/// </exception>
	public AesCbc(int size = 128)
	{
		if (Array.IndexOf(sizes, size) == -1)
			throw new ArgumentException("Key size must be 128 or 256 bits.");
		keySize = size;
	}

	/// <summary>Encrypts data (string)</summary>
	/// <param name="data">The data to encrypt.</param>
	/// <param name="password">The pasword.</param>
	/// <returns>Raw bytes or base64 encoded data.</returns>
	public byte[] Encrypt(string data, string password)
	{
		return Encrypt(Encoding.UTF8.GetBytes(data), password);
	}

	/// <summary>Encrypts data (bytes)</summary>
	public byte[] Encrypt(byte[] data, string password)
	{
		byte[] iv = IVGen();
		byte[] salt = IVGen();
		byte[][] keys = KeyGen(password, salt);
		key = keys[0];
		byte[] encrypted = Transformer(data, key, iv, true);

		List<byte> new_data = new List<byte>();
		new_data.AddRange(iv);
		new_data.AddRange(encrypted);
		byte[] hmac = Sign(new_data.ToArray(), keys[1]);
		new_data.InsertRange(0, salt);
		new_data.InsertRange(new_data.Count, hmac);

		if (this.b64)
			return Encoding.ASCII.GetBytes(Convert.ToBase64String(new_data.ToArray()));
		return new_data.ToArray();
	}

	/// <summary>Decrypts data (string)</summary>
	/// <param name="data">Encrypted data (base64 encoded)</param>
	/// <param name="password">The pasword.</param>
	/// <returns>Decrypted bytes.</returns>
	public byte[] Decrypt(string data, string password)
	{
		return Decrypt(Encoding.ASCII.GetBytes(data), password);
	}

	/// <summary>Decrypts data (bytes).</summary>
	public byte[] Decrypt(byte[] data, string password)
	{
		try
		{
			if (this.b64)
				data = Convert.FromBase64String((Encoding.ASCII.GetString(data)));
			if (data.Length < ivSize + ivSize + blockSize + macSize)
				throw new Exception("Not enough data.");

			List<byte> decoded = new List<byte>(data);
			byte[] salt = decoded.GetRange(0, ivSize).ToArray();
			byte[] iv = decoded.GetRange(ivSize, ivSize).ToArray();
			byte[] encrypted = decoded.GetRange(ivSize * 2, decoded.Count - (ivSize + ivSize + macSize)).ToArray();
			byte[] hmac = decoded.GetRange(decoded.Count - macSize, macSize).ToArray();
			byte[][] keys = KeyGen(password, salt);
			key = keys[0];
			byte[] iv_encrypted = decoded.GetRange(ivSize, decoded.Count - (ivSize + macSize)).ToArray();

			if (!Verify(iv_encrypted, hmac, keys[1]))
				throw new Exception("Verification failed.");
			byte[] decrypted = Transformer(encrypted, key, iv, false);
			return decrypted;
		}
		catch (Exception e)
		{
			Console.WriteLine(e.Message);
			return null;
		}
	}

	/// <summary>Encrypts / decrypts data.</summary>
	/// <param name="data">The data</param>
	/// <param name="key">The key</param>
	/// <param name="iv">The IV</param>
	/// <param name="encrypt">Flag</param>
	/// <returns>Eecrypted / ecrypted bytes.</returns>
	private byte[] Transformer(byte[] data, byte[] key, byte[] iv, bool encrypt)
	{
		using (Aes aes = Aes.Create())
		{
			aes.KeySize = keySize;
			aes.Mode = CipherMode.CBC;
			aes.Padding = PaddingMode.PKCS7;
			using (ICryptoTransform ict = (encrypt ? aes.CreateEncryptor(key, iv) : aes.CreateDecryptor(key, iv)))
			{
				using (MemoryStream mstream = new MemoryStream())
				{
					using (CryptoStream cstream = new CryptoStream(mstream, ict, CryptoStreamMode.Write))
					{
						cstream.Write(data, 0, data.Length);
					}
					return mstream.ToArray();
				}
			}
		}
	}

	/// <summary>Creates a pair of keys.</summary>
	/// <param name="password">The pasword.</param>
	/// <param name="salt">The salt.</param>
	/// <returns>keys</returns>
	private byte[][] KeyGen(string password, byte[] salt)
	{
		using (Rfc2898DeriveBytes kdf = new Rfc2898DeriveBytes(password, salt, rounds))
		{
			byte[] aesKey = kdf.GetBytes(keySize / 8);
			byte[] macKey = kdf.GetBytes(keySize / 8);
			return new byte[][] { aesKey, macKey };
		}
	}

	/// <summary>Creates random bytes.</summary>
	/// <param name="size">The number of bytes.</param>
	/// <returns>Random bytes.</returns>
	private byte[] IVGen(int size = 16)
	{
		byte[] iv = new byte[size];
		using (RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider())
		{
			rng.GetBytes(iv);
		}
		return iv;
	}

	/// <summary>Creates HMAC signature.</summary>
	/// <param name="data">The data.</param>
	/// <param name="key">The key.</param>
	/// <returns>HMAC</returns>
	private byte[] Sign(byte[] data, byte[] key)
	{
		using (HMACSHA256 hmac = new HMACSHA256(key))
		{
			return hmac.ComputeHash(data);
		}
	}

	/// <summary>Preforms HMAC verification.</summary>
	/// <param name="data">The data.</param>
	/// <param name="mac">The HMAC.</param>
	/// <param name="key">The key.</param>
	/// <returns>True if the HMACs match else false.</returns>
	private bool Verify(byte[] data, byte[] mac, byte[] key)
	{
		byte[] data_mac = Sign(data, key);
		for (int i = 0; i < data_mac.Length; i++)
		{
			if (data_mac[i] != mac[i])
				return false;
		}
		return true;
	}
}

